cli_tools/common/imagefile/qemu_img.go

```diff
@@ -53,34 +57,112 @@ func NewInfoClient() InfoClient {
 
 type defaultInfoClient struct{}
 
+type fileInfoJsonTemplate struct {
+	Filename         string `json:"filename"`
+	Format           string `json:"format"`
+	ActualSizeBytes  int64  `json:"actual-size"`
+	VirtualSizeBytes int64  `json:"virtual-size"`
+}
+
 func (client defaultInfoClient) GetInfo(ctx context.Context, filename string) (info ImageInfo, err error) {
 	if !files.Exists(filename) {
-		return info, fmt.Errorf("file %q not found", filename)
+		err = fmt.Errorf("file %q not found", filename)
+		return
+	}
+
+	jsonTemplate, err := client.getFileInfo(ctx, filename)
+	if err != nil {
+		err = daisy.Errf("Failed to inspect file %v: %v", filename, err)
+		return
+	}
+	info.Format = lookupFileFormat(jsonTemplate.Format)
+	info.ActualSizeBytes = jsonTemplate.ActualSizeBytes
+	info.VirtualSizeBytes = jsonTemplate.VirtualSizeBytes
+
+	checksum, err := client.getFileChecksum(ctx, filename, info.VirtualSizeBytes)
+	if err != nil {
+		err = daisy.Errf("Failed to calculate file '%v' checksum by qemu: %v", filename, err)
+		return
 	}
+
+	info.Checksum = checksum
+	return
+}
+
+func (client defaultInfoClient) getFileInfo(ctx context.Context, filename string) (*fileInfoJsonTemplate, error) {
 	cmd := exec.CommandContext(ctx, "qemu-img", "info", "--output=json", filename)
 	out, err := cmd.Output()
+	err = constructCmdErr(string(out), err, "inspection failure")
 	if err != nil {
-		var exitError *exec.ExitError
-		if errors.As(err, &exitError) {
-			return info, fmt.Errorf("inspection failure: %w, stderr: %s", err,
-				exitError.Stderr)
-		}
-		return info, fmt.Errorf("inspection failure: %w", err)
+		return nil, err
 	}
-	jsonTemplate := struct {
-		Filename         string `json:"filename"`
-		Format           string `json:"format"`
-		ActualSizeBytes  int64  `json:"actual-size"`
-		VirtualSizeBytes int64  `json:"virtual-size"`
-	}{}
+
+	jsonTemplate := fileInfoJsonTemplate{}
 	if err = json.Unmarshal(out, &jsonTemplate); err != nil {
-		return info, fmt.Errorf("failed to inspect %q: %w", filename, err)
+		return nil, daisy.Errf("failed to inspect %q: %w", filename, err)
+	}
+	return &jsonTemplate, err
+}
+
+func (client defaultInfoClient) getFileChecksum(ctx context.Context, filename string, virtualSizeBytes int64) (checksum string, err error) {
+	// Check size = 200000*512 = 100MB
+	checkCount := int64(200000)
+	blockSize := int64(512)
+	blockCount := virtualSizeBytes / blockSize
+	skips := []int64{0, int64(2000000) - checkCount, int64(20000000) - checkCount, blockCount - checkCount}
+	tmpOutFilePrefix := "out" + pathutils.RandString(5)
+	for i, skip := range skips {
+		tmpOutFileName := fmt.Sprintf("%v%v", tmpOutFilePrefix, i)
```

Comment:
It looks like test only checks if the md5sum is non-empty, so I think there's an opportunity to improve coverage.

Take a look at `ShellExecutor` which is a replacement for using `exec` directly.
 - Interface: https://github.com/GoogleCloudPlatform/compute-image-tools/blob/master/cli_tools/common/utils/shell/executor.go
 - Example usage: https://github.com/GoogleCloudPlatform/compute-image-tools/blob/c36c64a28b34f9459d83fbb3dac90f3e6955410e/cli_tools/common/mount/inspector_test.go#L27

Using these, you can verify the arguments passed to the shell command, and also verify that your go code correctly parses the response.

Reply:
Given that a bug in this checksum logic (either now or in the future) would cause an import to fail, I'd definitely be interested in having unit test coverage. We can talk as a team, though, as well. Maybe there's something I'm missing here!

Reply:
I modified the tests yesterday. The coverage is that I checked the checksum output follows the format "^[a-zA-Z0-9]{32}-[a-zA-Z0-9]{32}-[a-zA-Z0-9]{32}-[a-zA-Z0-9]{32}$". If the checksum calculation is failed, we won't get this format of output. Do you think that is a good coverage?

Reply:
done